### DISCUSSION TITLE
Griffe extension for Pydantic

### DISCUSSION THREAD

**Discussion by pawamoy** [pydantic/pydantic#6524] [Projects using Pydantic]
Hello team!

I'm working on a Griffe extension for Pydantic, and have a suggestion for you that would allow me to improve my results.

Take a look at this screenshot (yeah, teasing a bit :wink:):

![griffe_pydantic](https://github.com/pydantic/pydantic/assets/3999221/8eef2adf-8e1b-4120-ba41-33f7ba26d3cb)

(I'm completely getting inspiration from https://autodoc-pydantic.readthedocs.io, big thanks to @mansenfranzen!)

This is auto-generated by mkdocstrings, thanks to the Griffe extension that better understands Pydantic models and provides its own templates to render them.

Now, look at the "Config" closer: it's a simple list of `config_name: config_value`. What would be amazing is that we linked the config item names to their actual docs at https://docs.pydantic.dev. For this to be possible, you would have to document each item of the [ConfigDict](https://docs.pydantic.dev/dev-v2/api/config/#pydantic.config.ConfigDict) with a docstring below it (while keeping the Attributes table with just summaries of each item). This would render a heading (and permalink) for each item, in turn allowing the Griffe extension to automatically render links to the corresponding items: <code><a href="https://docs.pydantic.dev/dev-v2/api/config/#pydantic.config.ConfigDict.frozen">frozen</a>: False</code> (the `#pydantic.config.ConfigDict.frozen` anchor does not exist yet so it will just lead you to the top of the page).

To be clear, here's what I'm suggesting:

```python
class ConfigDict(TypedDict, total=False):
    """A dictionary-like class for configuring Pydantic models.

    Attributes:
        ...
        extra: Whether to ignore, allow, or forbid extra attributes during model initialization.
        frozen: Whether or not models are faux-immutable.
        ...
    """

    ...  # previous attributes
    extra: ExtraValues | None
    """Whether to ignore, allow, or forbid extra attributes during model initialization.
    Accepts the string values of `'ignore'`, `'allow'`, or `'forbid'`. Defaults to `'ignore'`.

    - `'forbid'` will cause validation to fail if extra attributes are included.
    - `'ignore'` will silently ignore any extra attributes.
    - `'allow'` will assign the attributes to the model.

    See [Extra Attributes](../usage/model_config.md#extra-attributes) for details.
    """
    frozen: bool
    """Whether or not models are faux-immutable, i.e. whether `__setattr__` is allowed, and also generates
    a `__hash__()` method for the model. This makes instances of the model potentially hashable if all the
    attributes are hashable. Defaults to `False`.

    !!! note
        On V1, this setting was called `allow_mutation`, and was `True` by default.
    """
    ...  # next attributes
```

Obviously, a few things are not ideal here: you have to repeat the summary (so have to be careful to keep things in sync), and there's no automatic link from your attributes table to the actual item headings. But good news, I'm planning on implementing such a feature: it would allow to auto-generate summary tables, for attributes, methods, whatever, using the summary of each item's docstring, with links going from the table to the actual headings.

But enough teasing!

This suggestion can of course apply to other parts of the source code / documentation, to make sure each item (attributes usually, since modules/classes/functions always have their own docstring) has a proper heading/permalink, so that we can link to it from other sites (giving a heading to an object means it's added to the inventory and can therefore be cross-ref'd) :slightly_smiling_face: 

Let me know what you think, I can open a PR with changes to the ConfigDict docstring if you want :smile:

**Comment by miili** [comment]
This looks good. What is the state on the griffe extension?

**Reply by pawamoy** [reply]
It's working, but is still very experimental and incomplete, and probably needs an update, to better use latest Griffe features. Also it's part of Insiders, so it's only available to sponsors :slightly_smiling_face:
